### PR TITLE
Enhancement for better deployment

### PR DESCRIPTION
**Pull request by lasofivec**
- Created a matrix for travis testing: now on mac, linux (trusty and xenial)
- Deployment only on master branch
- Deployment of wheel package if on mac
- Added skip_existing so that all build pass always
- Added possibility to test deployment (using source distribution only) if push to `test-deploy` branch. 

**Warning**: an account in test.pypi has to be created

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,69 +1,124 @@
 language: python
-python:
-- '3.6'
-- '3.7-dev'
 
-git:
-  depth: 200
```
Comment (lasofivec) [inline .travis.yml]: we don't go back to any old version of git so no need for this line

File: .travis.yml [hunk 2]
```diff
@@ -1,69 +1,124 @@
 language: python
-python:
-- '3.6'
-- '3.7-dev'
```
Comment (lasofivec) [inline .travis.yml]: 3.7 version now is available

File: .travis.yml [hunk 3]
```diff
@@ -1,69 +1,124 @@
...
-  nose-timer coverage codecov
-- export REV=$(python -c "import _updateversion as up; out=up.updateversion(); print(out)")
-- export VERSION=$(echo $REV | tr - .)
-- echo $REV
-- conda build conda_recipe
-- export PKG_DIR=$HOME/miniconda/conda-bld/linux-64/
-- conda install tofu --use-local
+  - if [[ "$TRAVIS_PYTHON_VERSION" == "3.7" ]]; then export VADD="py37";
+    else export VADD="py36";  fi
+  - wget "$REPO" -O miniconda.sh
```
Comment (lasofivec) [inline .travis.yml]: Change here since we created an environmental variable for the source of the conda repo

File: .travis.yml [hunk 4]
```diff
@@ -1,69 +1,124 @@
...
+  - cd $START
+
 deploy:
   - provider: script
     user: "ToFuProject"
     script: $START/anaconda_upload.sh
     on:
       tags: true
-      all_branches: true
+      branch: master
```
Comment (lasofivec) [inline .travis.yml]: no need to deploy for all branches only master

File: conda_recipe/conda_upload.sh [hunk 5]
```diff
@@ -1,7 +1,6 @@
 # Only need to change these two variables
 PKG_NAME=tofu
 USER=ToFuProject
-OS=linux-64
```
Comment (lasofivec) [inline conda_recipe/conda_upload.sh]: now this is was created in .travis.yml

File: conda_recipe/meta.yaml [hunk 6]
```diff
@@ -4,13 +4,11 @@ package:
...
 source:
   git_url: https://github.com/ToFuProject/tofu.git
-  #git_branch: {{ environ['TRAVIS_BRANCH'] }}
   git_rev: {{ environ['REV'] }}
-  # "patches:" might be the answer for clang compilers ?
 
 build:
   script_env:
-    #   - REV
+    - OS
```
Comment (lasofivec) [inline conda_recipe/meta.yaml]: we will use the variable from .travis.yml

File: conda_recipe/meta.yaml [hunk 7]
```diff
@@ -21,7 +19,7 @@ requirements:
   # here same as run, as we are using cython
   build:
     - python
-    - setuptools
+    - setuptools >=40.8.0
```
Comment (lasofivec) [inline conda_recipe/meta.yaml]: updated dependency

File: .travis.yml [hunk 8]
```diff
@@ -1,69 +1,124 @@
...
     skip_existing: true
+    skip_cleanup: true
     on:
       tags: true
-      all_branches: true
+      condition: $OS = osx-64
+      branch: master
+    password:
+      secure: JNEDTDJVx/2fXNfHntNQ99iDRNuQ4uB3y+DBWVIBycCT95+UCb36YPtKzmruEk/UUS29Xgq4IYCGdfCSWE9smKqG8tV1PcHiw705m+AzcpKy77YtzbVECFBxqY4W36O2pHrkwEUzP/7acjFwNsnUFzArqEzsBJ+KdLaa4OPHJXCh30GA0GyqlrXYbBKG+DA9hX5vtsGo4C6w9noALYF3fS7pKPiI6ipKFnAlzGgHQ7Ke0uQME8N3IAFhmh+Z5xMtIIDWxlnqv+KszdG4DIaGV/W6NIJNAbRhzkqUd+Chu6LoPAd/XkHDTeirR/MBkNUc5UcRJxRnP9rUTRo1gCO/buTYuNRgFkMvqhV5a033+x9edWgtUiKNJIMPLXOxe0RJvc5GWji+Co77HtHxRmGRM2rnYqWMtZeYZlFbUdvHu/8jf0d6I8jyUgAoJYdlMA2u/ipENP3S6by4epE9qycUPXiIVh6r3DZbf3vPTMFvTZYAjBrA0NOzihv1xgcXwemmNUFOQSpe0io4UcFxtS9lLMo+30UMQjCHSnbEVM3zSlZmbMOKpkVOlKlt8Lz5NxwVgWtu9FuW2pGukLtE8AWbqvY9urXAPZCQqZlOIklIjJQIqOITnuw9LEV09cgvPHXfdvNni3ldbMlIQ89zryM6dYvhYryTiEZGK4JDR3wAKJA=
+  - provider: pypi
```
Comment (Didou09) [inline .travis.yml]: Ok, I have created an account on test.pipy.org

File: conda_recipe/meta.yaml [hunk 9]
```diff
@@ -4,13 +4,11 @@ package:
 
 source:
   git_url: https://github.com/ToFuProject/tofu.git
-  #git_branch: {{ environ['TRAVIS_BRANCH'] }}
   git_rev: {{ environ['REV'] }}
-  # "patches:" might be the answer for clang compilers ?
 
 build:
   script_env:
-    #   - REV
```
Comment (Didou09) [inline conda_recipe/meta.yaml]: Are you sure we don't need REV ? it is used just above

Comment (lasofivec) [reply conda_recipe/meta.yaml]: it was commented, so i think no

Comment (Didou09) [reply conda_recipe/meta.yaml]: My mistake, ok

File: .travis.yml [hunk 10]
```diff
@@ -1,69 +1,121 @@
...
+    python: 3.6
+    env:
+      - REPO=https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
+      - OS=linux-64
+  - name: "osx python 3.7"
+    os: osx
+    language: generic
+    env:
+    - REPO=https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
+    - TRAVIS_PYTHON_VERSION=3.7
```
Comment (Didou09) [inline .travis.yml]: Variable TRAVIS_PYTHON_VERSION is not defined in all cases, is it normal ?

File: .travis.yml [hunk 11]
```diff
@@ -1,69 +1,121 @@
...
+  - conda update -q conda
+  - conda config --append channels conda-forge
+  - conda config --append channels tofuproject
+  - conda info -a
+  - conda install -q python="$TRAVIS_PYTHON_VERSION" conda-verify
+    nose nose-timer coverage codecov
+  - export REV=$(python -c "import _updateversion as up; out=up.updateversion(); print(out)")
+  - export VERSION=$(echo $REV | tr - .)
+  - echo $REV
+  - pip install -e ".[dev]"
```
Comment (Didou09) [inline .travis.yml]: If I understand correctly, we don't build a conda package anymore at this stage we just instal from local source using pip for the unti tests, right ?
Was pip included by default in miniconda ? (There is no line installing pip above)

File: .travis.yml (hunk 10 above)
Comment (lasofivec) [reply .travis.yml]: yes, it is defined when it's the default platform language. However on Mac it is not possible to define a job with default language "python" (thus the `language: generic`). 
See for example:

https://docs.travis-ci.com/user/multi-os/#python-example-unsupported-languages

File: .travis.yml (hunk 11 above)
Comment (lasofivec) [reply .travis.yml]: it is installed by default when you install python

**Comment by lasofivec** [comment]
all tests are passing with this branch ;)
However i'm testing the deployment to test-pypi, to be sure the deployment with master will work. I will update this PR when deployment to test-pypi is working.

**Comment by lasofivec** [comment]
Replaced by #328

**Pull request closed without merging**